Automattic/wp-calypso
# WPCOM SSH: Allow user to update an existing SSH key

#### Proposed Changes

Right now, when we delete an custom SSH key, the key is detached from all the sites.
So when the user adds a new one, it's needed to go and reassign the ssh key to each site again.

Let the user update the ssh key, solves this problem.

Below are some screenshots of this new functionality.

<img width="1173" alt="image" src="https://user-images.githubusercontent.com/36432/201983246-e9b9c1e2-d206-41e0-a61a-51b220d17644.png">

<img width="1112" alt="image" src="https://user-images.githubusercontent.com/36432/201982901-a25ad65c-80d7-417b-acbf-504004a16380.png">

#### Testing Instructions

##### Test 1

1. Make sure that you apply this patch to your sandbox: D91797-code
2. Sandbox public api.
3. Start calypso dev environment.
4. Navigate to `/me/security/ssh-key`
5. Add an ssh key.
6. Make sure that next to `Remove SSH Key` there is another button, named:`Update SSH Key`
7. Click that button
8. Make sure that a modal displaying `Are you sure you want to update this SSH key? It will affect all attached sites.` appears.
9. Press the OK button.
10. Make sure that a modal displaying the old ssh key and a textarea in order to enter a new one
   appears.
11. Enter a new SSH key and click `Update SSH Key`
12. Ensure that the button changes it's state.
13. After the new key is saved, ensure that the modal is auto closed and a banner indicating that
    the key was saved appears in the top right corner.

###### Test 2

1. Make sure that you apply this patch to your sandbox: D91797-code
2. Sandbox public api.
3. Start calypso dev environment.
4. Navigate to `/me/security/ssh-key`
5. Add an ssh key.
6. Make sure that next to `Remove SSH Key` there is another button, named:`Update SSH Key`
7. Click that button
8. Make sure that a modal displaying `Are you sure you want to update this SSH key? It will affect all attached sites.` appears.
9. Press the OK button.
10. Make sure that a modal displaying the old ssh key and a textarea in order to enter a new one
   appears.
11. Enter a new SSH key and click `Update SSH Key`
12. Ensure that the button changes it's state.
14. Close the modal before the ssh key is saved.
15. Ensure that the `Remove SSH Key button` is disabled.
16. Ensure that the `Update SSH Key` button can be pressed and you can view the modal.
17. Ensure that when the modal reopens, if the update is not yet completed, the textareas and the
    button are in the same state as when the modal was originally closed.

###### Test 3

1. Make sure that you apply this patch to your sandbox: D91797-code
2. Sandbox public api.
3. Start calypso dev environment.
4. Navigate to `/me/security/ssh-key`
5. Add an ssh key.
4. Switch your site to one that has SSH hosting enabled.
5. Navigate to hosting configuration  `/hosting-config/<your domain>` view
6. Make sure to attach the newly added SSH key to the site.
7. Click on your profile icon in the top right corner.
8. Click on `Security -> SSH Key`
9. Update the ssh key as described in `Test 1`
10. Click on `My Sites` on the top left corner
11. Click on `Settings -> Hosting Configuration`
12. Ensure that where previously it was attached the previous SSH key, now there is a loading
    indicator.
13. Wait till the loading is completed.
14. Ensure that the attached SSH key is not the previous one, but the updated SSH key instead


#### Pre-merge Checklist


Related to https://github.com/Automattic/dotcom-forge/issues/1135
